MicroInterpreter::tensors_size() always returns zero

@tensorflow/micro

**System information**
- Host OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Fedora 32
- TensorFlow installed from (source or binary): source
- Tensorflow version (commit SHA if source): 238981a91a9b780ab4449829469a71c5d668a273
- Target platform (e.g. Arm Mbed OS, Arduino Nano 33 etc.): x86

**Describe the problem**

The tensors_size() method of the MicroInterpreter class always returns zero.  

**Please provide the exact sequence of commands/steps when you ran into the problem**

To demonstrate, insert the following line:

`TF_LITE_MICRO_EXPECT_GT(static_cast<size_t>(0), interpreter.tensors_size());`

here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/micro_interpreter_test.cc#L87

And run:

`$ make -f tensorflow/lite/micro/tools/make/Makefile TARGET="x86" test_micro_interpreter_test`

I applied your suggestion and I get the following: 

```
Testing TestInterpreter
interpreter.tensors_size() > static_cast<size_t>(0) failed at tensorflow/lite/micro/micro_interpreter_test.cc:87
Output tensors not at index 0 are allocated from the persistent memory arena. Repeat calls will cause excess allocation!
Testing TestMultiTenantInterpreter
Testing TestKernelMemoryPlanning
Output tensors not at index 0 are allocated from the persistent memory arena. Repeat calls will cause excess allocation!
Output tensors not at index 0 are allocated from the persistent memory arena. Repeat calls will cause excess allocation!
Output tensors not at index 0 are allocated from the persistent memory arena. Repeat calls will cause excess allocation!
Output tensors not at index 0 are allocated from the persistent memory arena. Repeat calls will cause excess allocation!
Output tensors not at index 0 are allocated from the persistent memory arena. Repeat calls will cause excess allocation!
Output tensors not at index 0 are allocated from the persistent memory arena. Repeat calls will cause excess allocation!
Testing TestVariableTensorReset
Testing TestIncompleteInitialization
Testing InterpreterWithProfilerShouldProfileOps
Testing TestIncompleteInitializationAllocationsWithSmallArena
Failed to allocate tail memory. Requested: 240, available 208, missing: 32
Failed to allocate memory for context->eval_tensors, 240 bytes required
Failed starting model allocation.

Testing TestInterpreterDoesNotAllocateUntilInvoke
[RecordingMicroAllocator] Arena allocation total 768 bytes
[RecordingMicroAllocator] Arena allocation head 32 bytes
[RecordingMicroAllocator] Arena allocation tail 736 bytes
[RecordingMicroAllocator] 'TfLiteEvalTensor data' used 240 bytes with alignment overhead (requested 240 bytes for 10 allocations)
[RecordingMicroAllocator] 'Persistent TfLiteTensor data' used 0 bytes with alignment overhead (requested 0 bytes for 0 tensors)
[RecordingMicroAllocator] 'Persistent TfLiteTensor quantization data' used 0 bytes with alignment overhead (requested 0 bytes for 0 allocations)
[RecordingMicroAllocator] 'TfLiteTensor variable buffer data' used 40 bytes with alignment overhead (requested 12 bytes for 3 allocations)
[RecordingMicroAllocator] 'NodeAndRegistration struct' used 168 bytes with alignment overhead (requested 168 bytes for 3 NodeAndRegistration structs)
[RecordingMicroAllocator] 'Operator runtime data' used 0 bytes with alignment overhead (requested 0 bytes for 0 OpData structs)
7/8 tests passed
~~~SOME TESTS FAILED~~~

Interpreter has 0 tensors and 2 nodes
Inputs: 0
Outputs: 2 3


Node   0 Operator Custom Name mock_custom
  Inputs: 0 1
  Outputs: 2
Node   1 Operator Custom Name mock_custom
  Inputs: 0 1
  Outputs: 3
make: *** [tensorflow/lite/micro/tools/make/Makefile:368: test_micro_interpreter_test] Error 1
```

--- a/tensorflow/lite/micro/micro_interpreter.cc
+++ b/tensorflow/lite/micro/micro_interpreter.cc
@@ -274,6 +274,7 @@ TfLiteStatus MicroInterpreter::AllocateTensors() {
   // TODO(b/16157777): This call would not be needed if ContextHelper rolled
   // into the interpreter.
   context_helper_.SetTfLiteEvalTensors(eval_tensors_);
+  context_.tensors_size = subgraph_->tensors()->size();
 
   // If the system is big endian then convert weights from the flatbuffer from
   // little to big endian on startup so that it does not need to be done during

--- a/tensorflow/lite/micro/micro_interpreter_test.cc
+++ b/tensorflow/lite/micro/micro_interpreter_test.cc
@@ -270,7 +270,7 @@ TF_LITE_MICRO_TEST(TestVariableTensorReset) {
   tflite::AllOpsResolver op_resolver = tflite::testing::GetOpResolver();
 
   constexpr size_t allocator_buffer_size =
-      2096 /* optimal arena size at the time of writting. */ +
+      3072 /* optimal arena size at the time of writting. */ +
       16 /* alignment */ + 100 /* some headroom */;
   uint8_t allocator_buffer[allocator_buffer_size];
   tflite::MicroInterpreter interpreter(model, op_resolver, allocator_buffer,

--- a/tensorflow/lite/micro/micro_interpreter_test.cc
+++ b/tensorflow/lite/micro/micro_interpreter_test.cc
@@ -84,7 +84,7 @@ TF_LITE_MICRO_TEST(TestInterpreter) {
     TF_LITE_MICRO_EXPECT_LE(interpreter.arena_used_bytes(), 928 + 100);
     TF_LITE_MICRO_EXPECT_EQ(static_cast<size_t>(1), interpreter.inputs_size());
     TF_LITE_MICRO_EXPECT_EQ(static_cast<size_t>(2), interpreter.outputs_size());
-    TF_LITE_MICRO_EXPECT_GT(interpreter.tensors_size(), static_cast<size_t>(0));
+    TF_LITE_MICRO_EXPECT_GT(interpreter.tensors_size(), static_cast<size_t>(3));
 
     TfLiteTensor* input = interpreter.input(0);
     TF_LITE_MICRO_EXPECT_NE(nullptr, input);

--- a/tensorflow/lite/micro/micro_interpreter_test.cc
+++ b/tensorflow/lite/micro/micro_interpreter_test.cc
@@ -84,6 +84,7 @@ TF_LITE_MICRO_TEST(TestInterpreter) {
     TF_LITE_MICRO_EXPECT_LE(interpreter.arena_used_bytes(), 928 + 100);
     TF_LITE_MICRO_EXPECT_EQ(static_cast<size_t>(1), interpreter.inputs_size());
     TF_LITE_MICRO_EXPECT_EQ(static_cast<size_t>(2), interpreter.outputs_size());
+    TF_LITE_MICRO_EXPECT_GT(interpreter.tensors_size(), static_cast<size_t>(0));
 
     TfLiteTensor* input = interpreter.input(0);
     TF_LITE_MICRO_EXPECT_NE(nullptr, input);

--- a/tensorflow/lite/micro/micro_interpreter_test.cc
+++ b/tensorflow/lite/micro/micro_interpreter_test.cc
@@ -84,7 +84,7 @@ TF_LITE_MICRO_TEST(TestInterpreter) {
     TF_LITE_MICRO_EXPECT_LE(interpreter.arena_used_bytes(), 928 + 100);
     TF_LITE_MICRO_EXPECT_EQ(static_cast<size_t>(1), interpreter.inputs_size());
     TF_LITE_MICRO_EXPECT_EQ(static_cast<size_t>(2), interpreter.outputs_size());
-    TF_LITE_MICRO_EXPECT_GT(interpreter.tensors_size(), static_cast<size_t>(3));
+    TF_LITE_MICRO_EXPECT_EQ(static_cast<size_t>(4), interpreter.tensors_size());
 
     TfLiteTensor* input = interpreter.input(0);
     TF_LITE_MICRO_EXPECT_NE(nullptr, input);

